src/test/java/com/example/cinematicketbookingapp/controller/ReservationControllerIntegrationTest.java

```diff
@@ -0,0 +1,317 @@
+@AutoConfigureMockMvc
+@DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
+@ActiveProfiles("test")
+class ReservationControllerIntegrationTest {
+
+    @Autowired
+    MockMvc mockMvc;
+
+    @Autowired
+    ObjectMapper objectMapper;
+
+    @Autowired
+    private ReservationRepository reservationRepository;
+
+    @Autowired
+    private WebApplicationContext webApplicationContext;
+
+    @Autowired
+    private ScreeningRepository screeningRepository;
+
+    @Test
+    void webApplicationAndServletContextShouldBeLoadedProperlyForTests() {
+        ServletContext servletContext = webApplicationContext.getServletContext();
+
+        assertThat(servletContext).isNotNull();
+        assertTrue(servletContext instanceof MockServletContext);
+        assertThat(webApplicationContext.getBean(ControllerTestConstantValues.RESERVATION_CONTROLLER_BEAN_NAME)).isNotNull();
+    }
+
+    @Test
+    void createReservation_shouldOccurOptimisticLock_whenAtLeastTwoThreadsWantToReserveAtLeastOneTheSameSeatAtTheSameTime() throws JsonProcessingException, InterruptedException {
+        //given
+        ReservationCreationDataDto reservationCreationDataDto1 = ReservationCreationDataDto.builder()
+                .seatIds(Set.of(1L, 2L, 3L))
+                .screeningId(1L)
+                .numberOfAdultTickets(3)
+                .ownerName(ControllerTestConstantValues.CORRECT_USER_NAME)
+                .ownerSurname(ControllerTestConstantValues.CORRECT_USER_SURNAME)
+                .build();
+        ReservationCreationDataDto reservationCreationDataDto2 = ReservationCreationDataDto.builder()
+                .seatIds(Set.of(5L, 6L, 7L))
+                .screeningId(1L)
+                .numberOfAdultTickets(3)
+                .ownerName(ControllerTestConstantValues.CORRECT_USER_NAME)
+                .ownerSurname(ControllerTestConstantValues.CORRECT_USER_SURNAME)
+                .build();
+
+        String requestBody1 = objectMapper.writeValueAsString(reservationCreationDataDto1);
+        String requestBody2 = objectMapper.writeValueAsString(reservationCreationDataDto2);
+
+        MockHttpServletRequestBuilder requestWithOptimisticLock = MockMvcRequestBuilders.post(ControllerTestConstantValues.RESERVATIONS_PATH)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(requestBody1);
+        MockHttpServletRequestBuilder requestWithoutOptimisticLock = MockMvcRequestBuilders.post(ControllerTestConstantValues.RESERVATIONS_PATH)
+                .contentType(MediaType.APPLICATION_JSON)
+                .content(requestBody2);
+
+        ExecutorService executorService = Executors.newFixedThreadPool(4);
+
+        Callable callableWithOptimisticLock = () -> mockMvc.perform(requestWithOptimisticLock);
+        Callable callableWithoutOptimisticLock = () -> mockMvc.perform(requestWithoutOptimisticLock);
+
+        //when
+        executorService.submit(callableWithOptimisticLock); //reservation saved
```

Comment:
Co dzięki temu testujemy? Jaka jest szansa, że coś w tych testach uda się wykryć?

Reply:
Za pomocą tego testu sprawdzałem czy faktycznie lock działa, bo niestety manualnie nie miałem pojęcia jak to zrobić, a chciałem mieć pewność, że poprawnie zaimplementowałem rozwiązanie. Jest to mój własny pomysł i nie wiem czy w taki sposób i czy w ogóle się to testuje. Idea tego testu polegała na tym, że trzy oddzielne wątki strzelają do API requestem z chęcią rezerwacji tych samych miejsc i pożądanym rezultatem było pojawienie się tylko jednej rezerwacji w bazie danych. Test wykazał też, że możliwe jest zarezerwowanie w tym samym momencie miejsc o innym ID. Na trzy wątki chcące zarezerwować te same miejsca pojawiła się tylko jedna rezerwacja w bazie danych, a czwarty wątek zarezerwował miejsca bez problemu, ponieważ pytał bazę danych o miejsca z innym ID. Bez lock’a wszystkie cztery rezerwacje zostawały zapisane w bazie danych. Czy wyniki będą powtarzalne dla różnych środowisk? Tego nie wiem, ale na ponad kilkanaście uruchomień testów zawsze dostawałem wynik pozytywny. Bardziej chodziło mi o eksperyment, ale chętnie dowiedziałbym się czy to w ogóle się testuje i jeśli tak, to w jaki sposób zrobić to poprawnie?

Korzystając z okazji, że to już ostatni komentarz, na który odpowiadam, chciałbym podziękować za Code Review. Bardzo wartościowe jest kiedy ktoś rzuci inne spojrzenie na jakiś problem i zauważy rzeczy, o których czasami, chociażby z niewiedzy lub niedopatrzenia, bym nie pomyślał. Jeszcze raz dzięki i pozdrawiam serdecznie.

Reply:
Hmmm, skomentuję to tak.
1. Co do unikalności - ja bym wolał założyć constrainty na bazie. Jednak ktoś może tę adnotację przypadkiem usunąć albo strzelić "boczkiem", gdzie ten lock aplikacyjny nie będzie dostępny. A spójność danych ponad wszystko!
2. Takich rzeczy się raczej nie testuje, to jest ficzer frameworka i ufamy że działa.
3. Testy z wątkami są niedeterministyczne, ciężko jest tak napisać, żeby w 100% przetestowały to co mają przetestować.

Również dziękuję za Twoje odpowiedzi, bardzo merytoryczne!